### PR TITLE
ENH: Add MSTL algorithm for multi-seasonal time series decomposition

### PR DESCRIPTION
**Pull request by KishManani**
This PR:

- Adds an implementation of the [MSTL algorithm](https://arxiv.org/pdf/2107.13462.pdf) and already [exists in the Forecast package in R](https://www.rdocumentation.org/packages/forecast/versions/8.16/topics/mstl). This provides a method to decompose a time series which has multiple seasonal components. 
- Adds an example notebook to show how to use it.
- Modifies the DecomposeResults class to allow plotting of multiple seasonal components. 

There are two deviations from the original algorithm:

1) Missing data handling is omitted and left to the user to do beforehand as a pre-processing step. This is because an equivalent of the R method used in the original paper to impute missing values does not have a simple implementation in Python. I also don't think this is core to the algorithm.

2) The original paper also handled a case when there is no seasonal component and uses Friedman’s Super Smoother
function to just extract a trend. This implementation assumes that there is at least one seasonal component otherwise a user would not use a decomposition method like this. Once again I don't think this is core to the algorithm.

I still need to compile the documentation and add that. I will do that if this feature is considered useful and the overall approach is accepted. 

Also any feedback regarding improving the tests would be appreciated.

- [ ] closes #xxxx
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>

**Notes**:

* It is essential that you add a test when making code changes. Tests are not 
  needed for doc changes.
* When adding a new function, test values should usually be verified in another package (e.g., R/SAS/Stata).
* When fixing a bug, you must add a test that would produce the bug in main and
  then show that it is fixed with the new code.
* New code additions must be well formatted. Changes should pass flake8. If on Linux or OSX, you can
  verify you changes are well formatted by running 
  ```
  git diff upstream/main -u -- "*.py" | flake8 --diff --isolated
  ```
  assuming `flake8` is installed. This command is also available on Windows 
  using the Windows System for Linux once `flake8` is installed in the 
  local Linux environment. While passing this test is not required, it is good practice and it help 
  improve code quality in `statsmodels`.
* Docstring additions must render correctly, including escapes and LaTeX.

</details>

### CODE REVIEW

File: statsmodels/tsa/mstl.py
Comment (bashtage) [inline statsmodels/tsa/mstl.py:32]: I don't like **kwargs.  It would be better to use a specific parameter `stl_args` or `stl_kwargs` as a `dict[str,bool | int | None]`. The alternative would be to list each of the STL arguments as keyword only arguments, although there are a lot.

File: statsmodels/tsa/mstl.py
Comment (bashtage) [inline statsmodels/tsa/mstl.py:99]: I think many of these should be keyword only.  Not sure where to draw the line.  KW only make it a lot easier to make any future changes.

File: statsmodels/tsa/mstl.py
Comment (bashtage) [inline statsmodels/tsa/mstl.py:1]: Would be a good idea to reorg now with

tsa\
tsa\stl\
tsa\stl\__init__.py
tsa\stl\_stl.pyx
\tsa\stl\mstl.py

File: examples/notebooks/mstl_decomposition.ipynb
Comment (bashtage) [inline examples/notebooks/mstl_decomposition.ipynb:1]: Can you add an image for the examples directory and add this to the yaml file that lists notebooks?

**Review by bashtage — CHANGES_REQUESTED** [review]
Looks very close.  Some things that I think we should change, and then I'll give it another look.  Thanks.

**Comment by ChadFulton** [comment]
Thanks for the contribution, this looks great!  Is it possible to write some unit tests that check the values against what you would get from the paper and/or the R package?

**Comment by KishManani** [comment]
Hi @ChadFulton. 

The [paper](https://arxiv.org/pdf/2107.13462.pdf) used an electricity demand in Victoria dataset with a frequency of 30 mins. They then upsampled to hourly and limited their view to H1 of 2012.

I found the raw dataset here: https://github.com/tidyverts/tsibbledata/tree/master/data-raw/vic_elec/VIC2015. I applied the same transformations to get the data to the same time period and granularity. It looks pretty similar/identical to the paper. I've attached this dataset as `mstl_elec_vic.csv` file in this commit: [f42b54b](https://github.com/statsmodels/statsmodels/pull/8160/commits/f42b54b0d3319e46cc265829c83162a9a8c1dd03).

I then used this dataset in R with the following params: `res = mstl(x=y, iterate=2, s.window=c(11, 15), s.jump=1, t.jump=1, l.jump=1)` and saved this dataset as `mstl_test_results.csv` (naming convention chosen to follow what was done for STL). This dataset is also to be found in the same commit: [f42b54b](https://github.com/statsmodels/statsmodels/pull/8160/commits/f42b54b0d3319e46cc265829c83162a9a8c1dd03).

I've then added a test which shows that the results are pretty much identical. Let me know if you want me to add more!

I plan to update the example notebook with this more real world example and add that next.

**Comment by KishManani** [comment]
I have now updated the example notebook to include generating the electricity demand dataset (mentioned (here)[https://github.com/statsmodels/statsmodels/pull/8160#issuecomment-1057590807]) that is now used in the unit test*. I have also updated the image file. See [a38930d](https://github.com/statsmodels/statsmodels/pull/8160/commits/a38930d3559668d83eb669ecb34a5c5caced88c2).

* The dataset added in [091c8b5](https://github.com/statsmodels/statsmodels/pull/8160/commits/091c8b5194e299ae1fbe5927d275036332de1780) has one additional day of data relative to the notebook. However, this has no impact on the unit test.

File: statsmodels/tsa/stl/mstl.py
Comment (bashtage) [inline statsmodels/tsa/stl/mstl.py:126]: Somethign is wrong here. You first get `periods` and `windows`, and then right below you access `self.preiods`.  Should you be using these sorted periods below?  If so, could you add a test where you have periods out-of-order and also inorder and ensure the results are the same?

File: statsmodels/tsa/stl/mstl.py
Comment (bashtage) [inline statsmodels/tsa/stl/mstl.py:132]: Redefine without using the windows above.

File: statsmodels/tsa/stl/mstl.py
Comment (bashtage) [inline statsmodels/tsa/stl/mstl.py:130]: Redefine without using periods above.

File: statsmodels/tsa/stl/mstl.py
Comment (bashtage) [reply statsmodels/tsa/stl/mstl.py:126]: If you really need to sort, should you not just do this in the `__init__` so that there is a consistent canonical representation of models that have the same periods/windows in any order?

**Review by bashtage — CHANGES_REQUESTED** [review]
Probably a bug

File: statsmodels/tsa/stl/mstl.py
Comment (bashtage) [inline statsmodels/tsa/stl/mstl.py:26]: You should directly import STL from where it is defined, not from `seasonal` which is just an API end-point. This can produce cyclic import problems.

File: statsmodels/tsa/stl/mstl.py
Comment (KishManani) [reply statsmodels/tsa/stl/mstl.py]: The sorting logic is part of the algorithm which is why I was tempted to keep it in `.fit()`. But I think your suggestion also makes sense so I've refactored the logic for handling the various input types and the sorting of `windows` and `periods`, moved it to the `__init__`, and implemented  the suggested new test.

**Comment by KishManani** [comment]
@bashtage once the review is completed would you like me to squash my commits or is that done automatically when you merge? Or is squashing not done for contributing to Statsmodels? Thanks!

**Comment by bashtage** [comment]
@KishManani I'll leave it up to you.  We don't religiously squash before emerge.  I'm happy if you would like to go to 1 or leave it.  I think it looks good and I'll merge soon if there are no objections.

**Comment by KishManani** [comment]
@bashtage I'll leave it as it is then. Thank you!

**Comment by bashtage** [comment]
Thanks for this contribution.  It is great.

**Pull request merged**
